test/integration/lib/update-lock.spec.ts

```diff
@@ -365,7 +366,158 @@ describe('lib/update-lock', () => {
+			});
+
+			afterEach(async () => {
+				await testFs.restore();
+				await fs.rm(path.join(lockdir, '1'), { recursive: true });
+				await fs.rm(path.join(lockdir, '2'), { recursive: true });
+			});
+
+			it('takes locks for a list of services for an appId', async () => {
+				// Take locks for appId 1
+				await updateLock.takeLock(1, ['server', 'client']);
+				// Locks should have been taken
+				expect(lockfile.getLocksTaken()).to.deep.include.members(
+					serviceLockPaths[1],
+				);
+				expect(lockfile.getLocksTaken()).to.have.length(4);
+				expect(
+					await fs.readdir(path.join(lockdir, '1', 'server')),
+				).to.include.members(['updates.lock', 'resin-updates.lock']);
+				expect(
+					await fs.readdir(path.join(lockdir, '1', 'client')),
+				).to.include.members(['updates.lock', 'resin-updates.lock']);
+				// Take locks for appId 2
+				await updateLock.takeLock(2, ['main']);
+				// Locks should have been taken for appid 1 & 2
+				expect(lockfile.getLocksTaken()).to.deep.include.members([
+					...serviceLockPaths[1],
+					...serviceLockPaths[2],
+				]);
+				expect(lockfile.getLocksTaken()).to.have.length(6);
+				expect(
+					await fs.readdir(path.join(lockdir, '2', 'main')),
+				).to.have.length(2);
+				// Clean up the lockfiles
+				for (const lockPath of serviceLockPaths[1].concat(
+					serviceLockPaths[2],
+				)) {
+					await lockfile.unlock(lockPath);
+				}
+			});
+
+			it('creates lock directory recursively if it does not exist', async () => {
+				// Take locks for app with nonexistent service directories
+				await updateLock.takeLock(3, ['api']);
+				// Locks should have been taken
+				expect(lockfile.getLocksTaken()).to.deep.include(
+					path.join(lockdir, '3', 'api', 'updates.lock'),
+					path.join(lockdir, '3', 'api', 'resin-updates.lock'),
+				);
+				// Directories should have been created
+				expect(await fs.readdir(path.join(lockdir))).to.deep.include.members([
+					'3',
+				]);
+				expect(
+					await fs.readdir(path.join(lockdir, '3')),
+				).to.deep.include.members(['api']);
+				// Clean up the lockfiles & created directories
+				await lockfile.unlock(path.join(lockdir, '3', 'api', 'updates.lock'));
+				await lockfile.unlock(
+					path.join(lockdir, '3', 'api', 'resin-updates.lock'),
+				);
+				await fs.rm(path.join(lockdir, '3'), { recursive: true });
+			});
+
+			it('should not take lock for services where Supervisor-taken lock already exists', async () => {
+				// Take locks for one service of appId 1
+				await lockfile.lock(serviceLockPaths[1][0]);
+				await lockfile.lock(serviceLockPaths[1][1]);
+				// Sanity check that locks are taken & tracked by Supervisor
+				expect(lockfile.getLocksTaken()).to.deep.include(
+					serviceLockPaths[1][0],
+					serviceLockPaths[1][1],
+				);
+				expect(lockfile.getLocksTaken()).to.have.length(2);
+				// Take locks using takeLock, should not error
+				const lockSpy = sinon.spy(lockfile, 'lock');
```

Comment:
Can you think of a way to avoid using a spy here? Perhaps the `takeLock` function could return a list of the services for which locks have actually been taken?

Reply:
Removed the spy here: https://github.com/balena-os/balena-supervisor/pull/2234/commits/b56b0431f77aca675f1eb938c9be15110cc3c8cb

following your suggested solution.